GPU logical reshape pass-through

csullivan opened the pull request:
Add pass-through support for Reshape that doesn't change the memory layout (logical reshape only).

fengleitian commented on src/ngraph/runtime/gpu/gpu_emitter.cpp:
```diff
@@ -516,6 +516,36 @@ namespace ngraph
                     return;
                 }
                 auto reshape = static_cast<const op::Reshape*>(node);
+                auto can_skip_reshape = [&]() {
```
check if the input name is same as output name.

fengleitian replied:
if(same) 
   skip;
else 
   copy;

Krovatkin commented on src/ngraph/runtime/gpu/gpu_emitter.cpp:
```diff
@@ -516,6 +516,36 @@ namespace ngraph
                 }
                 auto reshape = static_cast<const op::Reshape*>(node);
+                auto can_skip_reshape = [&]() {
+                    if (!reshape->get_is_transpose())
```
Please don't do this! Emitter isn't a MemoryLayout pass. The decision whether to copy or not should be as localized as possible.
Do this:
```cpp
if (args[0].get_name() == out[0].get_name())
{
   // print skipping comment
}
else
{
  // emit a copy
}
```

fengleitian replied:
@Krovatkin 
anything to do with `op_annotations->add_in_place_oi_pair({0, 0, false});`? should we use true here?

Krovatkin replied:
Logical `Reshape`s are non-destructive, so it should be `false`. It's `true` for `ReplaceSlice`, for example. The reason why CPU has those checks in Emitter is somewhat convoluted and it has to do with a `CPULayout` pass optimizing away unnecessary transposes. You guys shouldn't have these issues so a simple equal comparison should work.

csullivan commented:
@rkimballn1 @fengleitian , the problem I was seeing with the logical reshape elimination was coming from the CommonFunctionCollection pass. By caching the function emission for pass-through ops the elimination of the reshape wouldn't end up being emitted, because the version with the memcpy had already been cached. I've added a check in CommonFunctionCollection to not cache ops with `oi_pairs` and simplified the reshape emitter check.

I've also added functions to propagate in place inputs and outputs - borrowed from IA. Thanks to @Krovatkin for pointing them out.

fengleitian commented on src/ngraph/runtime/gpu/pass/gpu_layout.cpp:
```diff
@@ -54,6 +55,33 @@ namespace ngraph
+                    auto reshape = static_cast<ngraph::op::Reshape*>(node.get());
+                    if (reshape->get_is_transpose() &&
+                        reshape->get_output_shape().size() ==
+                            reshape->get_argument(0)->get_shape().size())
```
does `reshape->get_output_shape().size() ==
                             reshape->get_argument(0)->get_shape().size()` required. Looks fine if we go 1^m to 1^n, or {2, 2} to {4} etc.

fengleitian replied:
Actually we don't need to check if it's identical.

fengleitian replied:
take it back, here you are check the case that can not do inplace.

fengleitian commented on src/ngraph/runtime/gpu/gpu_emitter.cpp:
```diff
@@ -516,17 +516,23 @@ namespace ngraph
                 }
                 auto reshape = static_cast<const op::Reshape*>(node);
+
+                if (out[0].get_name() == args[0].get_name() && !reshape->get_is_transpose())
```
since we don’t have a in place transpose.  `out[0].get_name() == args[0].get_name()` should covered everything.

csullivan replied:
If we add one we will need to remember to put this back. Does it hurt anything having it here?

csullivan replied:
All right, we can leave it for later.
